atomic_reactor/dirs.py

```diff
@@ -0,0 +1,227 @@
+        for platform in self.platforms:
+            copytree(src_path, self.path / platform.value, ignore=ignore)
+
+    @property
+    def has_sources(self) -> bool:
+        """Check if all platform-specific directories exist.
+
+        :return: True if all platform-specific directories exist. Otherwise,
+            False is returned.
+        :rtype: bool
+        """
+        for platform in self.platforms:
+            if not self.path.joinpath(platform.value).exists():
+                return False
+        return True
+
+    @property
+    def any_build_dir(self) -> BuildDir:
+        """Get a platform-specific build directory.
+
+        This is typically used by code that does not care about platform. The
+        returned directory is guaranteed to be the same one between the
+        multiple calls.
+
+        :return: a source directory.
+        :rtype: BuildDir
+        """
+        platform: Platform = sorted(self.platforms, key=lambda item: item.value)[0]
+        return BuildDir(self.path / platform.value, platform)
+
+    def for_each(self, action: Callable[[BuildDir], Any]) -> Dict[Platform, Any]:
+        """Apply an action on every platform-specific directory.
+
+        The action callable will be applied to the platform-specific
+        directories in the order of the platforms specified when initializing
+        this RootBuildDir object. The caller should not depend on or assume the
+        order how action is applied.
+
+        The action callable might raise errors, that will cause the
+        ``for_each`` terminates immediately and the raised error is propagated
+        to the caller. As a result, the action will not be applied to the rest
+        of the platforms.
+
+        :param action: a callable object that will be applied on every
+            platform-specific directory. This callable must accept one single
+            argument in BuildDir type, and it can return data in any type.
+        :type action: Callable
+        :return: a mapping from platform to the value returned from the
+            function which is called for that platform.
+        :rtype: dict[Platform, any]
+        """
+        results = {}
+        for platform in self.platforms:
+            build_dir = BuildDir(self.path / platform.value, platform)
+            results[platform] = action(build_dir)
+        return results
+
+    def for_all_copy(self, action: Callable[[BuildDir], Iterable[Path]]) -> List[Path]:
+        """Ensure created files are present in all platform-specific directories.
+
+        :param action: a callable that creates files in a given build directory
+            and returns the files it created. The action accepts one single
+            argument in type BuildDir, and returns an iterable object that
+            yields paths of the created files.
+        :type action: callable
+        :return: the list of absolute paths of the created files in all build
+            directories.
```

Comment:
I'd say this should return the files for just one build directory - they are guaranteed to be identical anyway